fog/ledger/enclave/impl/src/key_image_store.rs

```diff
@@ -0,0 +1,198 @@
+    pub fn new(desired_capacity: u64, logger: Logger) -> Self {
+        Self {
+            omap: Box::new(<ObliviousMapCreator<OSC> as OMapCreator<
+                KeySize,
+                ValueSize,
+                McRng,
+            >>::create(
+                desired_capacity, STASH_SIZE, McRng::default
+            )),
+            logger,
+        }
+    }
+
+    /// add a key image containing block index and timestamp
+    pub fn add_record(
+        &mut self,
+        key_image: &KeyImage,
+        block_index: u64,
+        timestamp: u64,
+    ) -> Result<(), AddRecordsError> {
+        let mut value = A8Bytes::<ValueSize>::default();
+        let mut key = A8Bytes::<KeySize>::default(); // key used to add to the oram for key image
+        key.clone_from_slice(&key_image.as_ref());
+        // Flip the first byte of key image, when used as a key in the oblivious
+        // map. This is because we will use key image as a key in the map,
+        // but the map does not support all zeroes as a key. All zeroes is a
+        // valid curve point. But if we flip the first byte, it turns out that
+        // isn't a valid curve point, in the Ristretto group.
+        // So this prevents the OMAP_INVALID_KEY error path.
+        key[0] = !key[0];
+        // write block index data to  value[0..8] write the time stamp data to
+        // value[8..16]
+        value[0..8].clone_from_slice(&block_index.to_le_bytes());
+        value[8..16].clone_from_slice(&timestamp.to_le_bytes());
+        // Note: Passing true means we allow overwrite, which seems fine since
+        // the search_key value is not changing
+        let omap_result_code = self.omap.vartime_write(&key, &value, Choice::from(1));
+        if omap_result_code == OMAP_INVALID_KEY {
+            return Err(AddRecordsError::KeyRejected);
+        } else if omap_result_code == OMAP_OVERFLOW {
+            return Err(AddRecordsError::MapOverflow(
+                self.omap.len(),
+                self.omap.capacity(),
+            ));
+        } else if omap_result_code == OMAP_FOUND {
+            log::debug!(self.logger, "A key image record was clobbered");
+        } else if omap_result_code != OMAP_NOT_FOUND {
+            panic!(
+                "omap_result_code had an unexpected value: {}",
+                omap_result_code
+            );
+        }
+        Ok(())
+    }
+
+    /// return new struct KeyImageResult which contains block index and
+    /// timestamp of key image as ref to convert key image to 32 bits,
+    /// call the oram to query to to key image data
+    pub fn find_record(&mut self, key_image: &KeyImage) -> KeyImageResult {
+        let mut result = KeyImageResult {
+            key_image: *key_image,
+            spent_at: u64::MAX,
+            key_image_result_code: KeyImageResultCode::KeyImageError as u32,
+            timestamp: u64::MAX,
+            timestamp_result_code: TimestampResultCode::TimestampFound as u32,
```

Comment:
this is a good question

the API currently documents like this: https://github.com/mobilecoinfoundation/fog/blob/30ec81169f96f3760694b8a01a1411beed85b187/fog/api/proto/ledger.proto#L126

```
/// The value is u64::MAX if the timestamp cannot be found. Callers must check the
/// timestamp_result_code to determine why the timestamp could not be found.
```

the result code is documented like this:

```
    /// The result code indicating whether the timestamp was found, can be tried again later, or will
    /// never be found with the current configuration of the Key Image Service's watcher.
    /// This is fixed32 to avoid leaking information about found / not found in the size of the encrypted
    /// payload.
    /// The possible values are described in enum TimestampResultCode.
    fixed32 timestamp_result_code = 4;
```

The behavior before this commit is that, when a client makes a key image check, we hit ledger db and watcher db, and return both codes basically. So I think in the origin block, you would get `Unavailable`, as you are saying.

After this commit, we only add key images to the ORAM when the timestamp is found, and if the timestamp isn't found, the loading thread is blocked. The clients no longer have to retry if they can't get a timestamp the first time.

So because of that I think the right answer is always TimestampFound, EXCEPT possibly the origin block, where the answer would have been `u64::MAX` (by convention) and `Unavailable`.

However, there are not any key images in the origin block, so that scenario actually never happens.

In a future commit, I would kind of like to eliminate the TimestampResultCode from this API, because all the watcher errors are handled in the server side at this point, and the clients don't need to have retries if they get a timestamp error anymore. We could tell the SDKs at some point that we are just going to set this to zero to save bandwidth and their code needs to be okay with that. For now I think it's okay to just always set this to `TimestampFound`, lmk if this makes sense to you.

Reply:
It's possible that we should update the comments in this proto in this commit to reflect this change, to make it clear that none of the other codes are going to be used

Reply:
Thanks for the detailed explanation @garbageslam 
Might be worth mentioning some of this in a comment.

Reply:
added a comment none of other codes will be used in proto